src/python/pants/backend/python/tasks/build_local_python_distributions.py

```diff
@@ -0,0 +1,88 @@
+# coding=utf-8
+# Copyright 2017 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import glob
+import os
+import shutil
+
+from pex.interpreter import PythonInterpreter
+
+from pants.backend.python.tasks.pex_build_util import is_local_python_dist
+from pants.backend.python.tasks.setup_py import SetupPyRunner
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TargetDefinitionException, TaskError
+from pants.base.fingerprint_strategy import DefaultFingerprintStrategy
+from pants.task.task import Task
+from pants.util.dirutil import safe_mkdir
+
+
+class BuildLocalPythonDistributions(Task):
+  """Create python distributions (.whl) from python_dist targets."""
+
+  options_scope = 'python-create-distributions'
+  PYTHON_DISTS = 'user_defined_python_dists'
+
+  @classmethod
+  def product_types(cls):
+    return [cls.PYTHON_DISTS]
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    round_manager.require_data(PythonInterpreter)
+
+  @property
+  def cache_target_dirs(self):
+    return True
+
+  def execute(self):
+    dist_targets = self.context.targets(is_local_python_dist)
+    built_dists = set()
+
+    if dist_targets:
+      with self.invalidated(dist_targets,
+                            fingerprint_strategy=DefaultFingerprintStrategy(),
+                            invalidate_dependents=True) as invalidation_check:
+        for vt in invalidation_check.all_vts:
+          if vt.valid:
+            built_dists.add(self._get_whl_from_dir(os.path.join(vt.results_dir, 'dist')))
+          else:
+            if vt.target.dependencies :
+              raise TargetDefinitionException(
+                vt.target, 'The `dependencies` field is disallowed on `python_dist` targets. List any 3rd '
+                           'party requirements in the install_requirements argument of your setup function.'
+              )
```

Comment:
AFAICT, this is not possible in the __init__ method. My attempts have been blocked by `AssertionError: Cannot retrieve dependencies of BuildFileAddress(testprojects/src/python/python_distribution/fasthello_with_install_requires/BUILD, fasthello) because it is not in the BuildGraph.`

Reply:
Yea, not feasible elsewhere: target constructors don't see their dependencies currently.

Reply:
the internal `Target.dependencies` representation isn't necessary for this check, afaik - the target definition in the BUILD file (`python_dist(arg=1, other_arg=2)`) maps 1:1 to `PythonDist.__init__(arg=1, other_arg=2)` via the symbol table as executed during BUILD file parsing - so theoretically all you need is e.g. `assert 'dependencies' not in kwargs` before the `super()` call.

Reply:
ah, k

Reply:
✅